aptos-move/framework/supra-framework/sources/multisig_voting.move

```diff
@@ -0,0 +1,1327 @@
+                },
+            );
+        };
+        event::emit_event<CreateProposalEvent>(
+            &mut voting_forum.events.create_proposal_events,
+            CreateProposalEvent {
+                proposal_id,
+                execution_hash,
+                expiration_secs,
+                metadata,
+                min_vote_threshold,
+            },
+        );
+
+        proposal_id
+    }
+
+    /// Vote on the given proposal.
+    ///
+    /// @param _proof Required so only the governance module that defines ProposalType can initiate voting.
+    /// @param voter signer of the voter.
+    /// @param voting_forum_address The address of the forum where the proposals are stored.
+    /// @param proposal_id The proposal id.
+    /// @param should_pass Whether the votes are for yes or no.
+    public fun vote<ProposalType: store>(
+        _proof: &ProposalType,
+        voter: &signer,
+        voting_forum_address: address,
+        proposal_id: u64,
+        should_pass: bool,
+    ) acquires VotingForum {
+        let voter_address = signer::address_of(voter);
+        let voting_forum = borrow_global_mut<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = table::borrow_mut(&mut voting_forum.proposals, proposal_id);
+
+        assert!(vector::contains(&proposal.voters, &voter_address), error::permission_denied(ENOT_MULTISIG_OWNER));
+
+        // Voting might still be possible after the proposal has enough yes votes to be resolved early. This would only
+        // lead to possible proposal resolution failure if the resolve early threshold is not definitive (e.g. < 50% + 1
+        // of the total voting token's supply). In this case, more voting might actually still be desirable.
+        // Governance mechanisms built on this voting module can apply additional rules on when voting is closed as
+        // appropriate.
+        assert!(!is_voting_period_over(proposal), error::invalid_state(EPROPOSAL_VOTING_ALREADY_ENDED));
+        assert!(!proposal.is_resolved, error::invalid_state(EPROPOSAL_ALREADY_RESOLVED));
+        // Assert this proposal is single-step, or if the proposal is multi-step, it is not in execution yet.
+        assert!(!simple_map::contains_key(&proposal.metadata, &utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY))
+            || *simple_map::borrow(&proposal.metadata, &utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY)) == to_bytes(
+            &false
+        ),
+            error::invalid_state(EMULTI_STEP_PROPOSAL_IN_EXECUTION));
+
+        if (table::contains(&proposal.voted_records, voter_address)) {
+            let voted_record = table::borrow_mut(&mut proposal.voted_records, voter_address);
+            assert!(voted_record != &should_pass, error::already_exists(EVOTE_DUPLICATE_VOTE));
+            *voted_record = should_pass;
+
+            if (should_pass) {
+                proposal.yes_votes = proposal.yes_votes + 1;
+                proposal.no_votes = proposal.no_votes - 1;
```

Comment:
```suggestion
 // Since the entry existed in `voted_records`, the vote flip should be reflected by this subtraction
                proposal.no_votes = proposal.no_votes - 1;
```